File: day08/day08.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Box struct {
	x, y, z int64
}

type Pair struct {
	distSq int64
	i, j   int
}

type UnionFind struct {
	parent     []int
	rank       []int
	components int
}

func NewUnionFind(n int) *UnionFind {
	parent := make([]int, n)
	for i := range parent {
		parent[i] = i
	}
	return &UnionFind{
		parent:     parent,
		rank:       make([]int, n),
		components: n,
	}
}

func (uf *UnionFind) Find(x int) int {
	if uf.parent[x] != x {
		uf.parent[x] = uf.Find(uf.parent[x]) // Path compression
	}
	return uf.parent[x]
}

func (uf *UnionFind) Union(x, y int) bool {
	px, py := uf.Find(x), uf.Find(y)
	if px == py {
		return false
	}

	// Union by rank
	if uf.rank[px] < uf.rank[py] {
		uf.parent[px] = py
	} else if uf.rank[px] > uf.rank[py] {
		uf.parent[py] = px
	} else {
		uf.parent[py] = px
		uf.rank[px]++
	}
	uf.components--
	return true
}

func (uf *UnionFind) CircuitSizes() []int {
	counts := make(map[int]int)
	for i := range uf.parent {
		counts[uf.Find(i)]++
	}

	sizes := make([]int, 0, len(counts))
	for _, size := range counts {
		sizes = append(sizes, size)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(sizes)))
	return sizes
}

func distSq(a, b Box) int64 {
	dx := b.x - a.x
	dy := b.y - a.y
	dz := b.z - a.z
	return dx*dx + dy*dy + dz*dz
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to open input.txt: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	var boxes []Box
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		parts := strings.Split(line, ",")
		x, _ := strconv.ParseInt(parts[0], 10, 64)
		y, _ := strconv.ParseInt(parts[1], 10, 64)
		z, _ := strconv.ParseInt(parts[2], 10, 64)
		boxes = append(boxes, Box{x, y, z})
	}

	n := len(boxes)

	// Generate all pairs sorted by distance
	var pairs []Pair
	for i := 0; i < n-1; i++ {
		for j := i + 1; j < n; j++ {
			pairs = append(pairs, Pair{distSq(boxes[i], boxes[j]), i, j})
		}
	}
	sort.Slice(pairs, func(a, b int) bool {
		return pairs[a].distSq < pairs[b].distSq
	})

	// Part 1: Connect 1000 shortest pairs
	uf1 := NewUnionFind(n)
	for i := 0; i < 1000; i++ {
		uf1.Union(pairs[i].i, pairs[i].j)
	}

	sizes := uf1.CircuitSizes()
	part1 := int64(sizes[0]) * int64(sizes[1]) * int64(sizes[2])

	// Part 2: Find last connection that unifies all circuits
	uf2 := NewUnionFind(n)
	var lastA, lastB int
	for _, p := range pairs {
		if uf2.Union(p.i, p.j) && uf2.components == 1 {
			lastA, lastB = p.i, p.j
			break
		}
	}

	part2 := boxes[lastA].x * boxes[lastB].x

	fmt.Printf("Part 1: %d\n", part1)
	fmt.Printf("Part 2: %d\n", part2)
}

File: day08/Day08.cs
var boxes = File.ReadAllLines("input.txt")
    .Select(line => line.Split(','))
    .Where(parts => parts.Length == 3)
    .Select(parts => (X: long.Parse(parts[0]), Y: long.Parse(parts[1]), Z: long.Parse(parts[2])))
    .ToArray();

var n = boxes.Length;

long DistSq((long X, long Y, long Z) a, (long X, long Y, long Z) b)
{
    var dx = b.X - a.X;
    var dy = b.Y - a.Y;
    var dz = b.Z - a.Z;
    return dx * dx + dy * dy + dz * dz;
}

// All pairs sorted by distance
var pairs = new List<(long Dist, int I, int J)>();
for (var i = 0; i < n - 1; i++)
    for (var j = i + 1; j < n; j++)
        pairs.Add((DistSq(boxes[i], boxes[j]), i, j));

pairs.Sort((a, b) => a.Dist.CompareTo(b.Dist));

// Union-Find with path compression and union by rank
int[] parent = Enumerable.Range(0, n).ToArray();
int[] rank = new int[n];
int components = n;

int Find(int x)
{
    if (parent[x] != x)
        parent[x] = Find(parent[x]);
    return parent[x];
}

bool Union(int x, int y)
{
    var px = Find(x);
    var py = Find(y);
    if (px == py) return false;

    if (rank[px] < rank[py])
        parent[px] = py;
    else if (rank[px] > rank[py])
        parent[py] = px;
    else
    {
        parent[py] = px;
        rank[px]++;
    }
    components--;
    return true;
}

int[] CircuitSizes()
{
    return Enumerable.Range(0, n)
        .GroupBy(Find)
        .Select(g => g.Count())
        .OrderByDescending(x => x)
        .ToArray();
}

// Part 1: Connect 1000 shortest pairs
for (var i = 0; i < 1000; i++)
    Union(pairs[i].I, pairs[i].J);

var sizes = CircuitSizes();
var part1 = (long)sizes[0] * sizes[1] * sizes[2];

// Reset for Part 2
parent = Enumerable.Range(0, n).ToArray();
rank = new int[n];
components = n;

// Part 2: Find last connection that unifies all circuits
(int A, int B) lastPair = (0, 0);
foreach (var (_, a, b) in pairs)
{
    if (Union(a, b) && components == 1)
    {
        lastPair = (a, b);
        break;
    }
}

var part2 = boxes[lastPair.A].X * boxes[lastPair.B].X;

Console.WriteLine($"Part 1: {part1}");
Console.WriteLine($"Part 2: {part2}");
